theano/gof/wrapper.py

```diff
@@ -0,0 +1,446 @@
+"""
+Module for wrapping many Theano variables into one struct for param ops.
+
+This module contains two classes:
+    - Wrapper: class to define the param op type.
+    - Wrap: convenient class to create an object that is compatible with the param op type.
+
+Example of usage:
+    # Importation
+        >>> from theano.common import Wrapper, Wrap
+    # In a op you create:
+        >>> params_type = Wrapper(attr1=TensorType('int32', (False, False)), attr2=TensorType('float64', (True,False)))
+    # In the get_params() method of your op:
+        >>> return Wrap(attr1=numpyArray1, attr2=numpyArray2)
+    # In perform() implementation (with params named `param`):
+        >>> print(param.attr1)
+        >>> print(param.attr2)
+    # In c_code() implementation (with `param = sub['params']`):
+        ```
+        PyArrayObject* attr1 = param.attr1;
+        PyArrayObject* attr2 = param.attr2;
+        /* Just use attr1 and attr2, you won't need to free them or whatever else. */
+        ```
+
+See theano/common/tests/test_wrapper.py for a complete working example.
+
+"""
+
+from __future__ import absolute_import, print_function, division
+import re
+import hashlib
+import numpy
+from theano.gof.utils import MethodNotDefined
+from theano.gof import Type
+from theano.gof.cmodule import GCC_compiler as compiler
+from theano.tensor.utils import hash_from_ndarray
+
+# NB: Maybe we should check if an attribute name is a C/C++ keyword, and raise an error if so.
+# These are some lists of C/C++ keywords:
+# http://fr.cppreference.com/w/cpp/keyword
+# http://fr.cppreference.com/w/c/keyword
+
+
+class Wrap(object):
+    """
+    Internal convenient class to wrap many Python objects into one
+    (this class is not safe as the hash method does not check if values are effectively hashable).
+
+    Example:
+        >>> w = Wrap(attr1=var1, attr2=var2, attri=vari)
+        >>> print(w.attr1, w.attr2, w.attri)
+        >>> d = dict(a=1, b=2, c='test')
+        >>> w2 = Wrap(**d)
+        >>> print(w2.a, w2.b, w2.c)
+
+    """
+
+    def __init__(self, **kwargs):
+        if len(kwargs) == 0:
+            raise TypeError('Wrap: cannot wrap empty data.')
+        # We want to use only the params provided in kwargs to hash the object,
+        # so I prefer to put them into a separate attribute (self.data) instead
+        # of directly in self.__dict__, to avoid confusion with builtin fields.
+        super(Wrap, self).__setattr__('data', kwargs)
```

Comment:
In this case you could consider making Wrap inherit from dict so that we have a separate dict, but also so that access to the values from C is faster and doesn't go through python.

But this is not a blocker.